homeassistant/components/onkyo/media_player.py

```diff
@@ -471,64 +482,66 @@ async def async_select_output(self, output):
         self._update_avr("hdmi-output-selector", output)
 
     @callback
-    def _parse_source(self, source_raw):
-        values = _parse_onkyo_tuple(source_raw)
-
-        for source in values:
+    def _parse_source(self, source):
+        # source is either a tuple of values or a single value,
+        # so we determine the source based on the type of source
+        if isinstance(source, str):
             if source in self._source_mapping:
                 self._source = self._source_mapping[source]
-                break
-            self._source = "_".join(values)
+            else:
+                self._source = source
+        else:
+            for value in source:
+                if value in self._source_mapping:
+                    self._source = self._source_mapping[value]
+                    break
+                self._source = "_".join(source)
 
     @callback
-    def _parse_sound_mode(self, sound_mode_raw):
-        values = _parse_onkyo_tuple(sound_mode_raw)
-
+    def _parse_sound_mode(self, sound_mode):
         # If the selected sound mode is not available, N/A is returned
         # so only update the sound mode when it is not N/A
-        if "N/A" not in values:
-            for sound_mode in values:
-                if sound_mode in SOUND_MODE_REVERSE_MAPPING:
-                    self._sound_mode = SOUND_MODE_REVERSE_MAPPING[sound_mode]
-                    break
-                self._sound_mode = "_".join(values)
+        if sound_mode != "N/A":
+            if sound_mode in SOUND_MODE_REVERSE_MAPPING:
+                self._sound_mode = SOUND_MODE_REVERSE_MAPPING[sound_mode]
+            else:
+                self._sound_mode = sound_mode
 
     @callback
-    def _parse_audio_inforamtion(self, audio_information_raw):
-        values = _parse_onkyo_tuple(audio_information_raw)
-        if "N/A" not in values:
-            info = {
-                "format": _tuple_get(values, 1),
-                "input_frequency": _tuple_get(values, 2),
-                "input_channels": _tuple_get(values, 3),
-                "listening_mode": _tuple_get(values, 4),
-                "output_channels": _tuple_get(values, 5),
-                "output_frequency": _tuple_get(values, 6),
-            }
+    def _parse_audio_inforamtion(self, audio_information):
+        # If audio information is not available, N/A is returned
+        # so only update the audio information when it is not N/A
+        if audio_information != "N/A":
```

Comment:
How about:
```
AUDIO_INFORMATION_MAPPING = [
    "audio_input_port",
    ....
]

...

if audio_information != "N/A":
    self._attributes.pop(ATTR_AUDIO_INFORMATION, None)
    return

info = {name: value for name, value in zip(AUDIO_INFORMATION_MAPPING, audio_information)}
self._attributes[ATTR_AUDIO_INFORMATION] = info
```

Reply:
With this implementation, I miss the check if the value length is greater than 0.
It now also fills in the empty parameters: 

```
audio_information: 
audio_input_port: NETWORK
input_signal_format: ''
input_frequency: ''
input_channels: ''
listening_mode: Extended Stereo
output_channels: 5.1.4 ch
output_frequency: ''
precision_quartz_lock_system: 'Off'
auto_phase_control_delay: '---'
auto_phase_control_phase: Normal
```